gpytorch/lazy/triangular_lazy_tensor.py

```diff
@@ -0,0 +1,101 @@
+
+
+class TriangularLazyTensor(LazyTensor):
+    def __init__(self, tensor, upper=False):
+        """
+        Triangular lazy tensor. Supports arbitrary batch sizes.
+
+        Args:
+            :attr:`tensor` (Tensor or LazyTensor):
+                A `b1 x ... x bk x n x n` Tensor, representing a `b1 x ... x bk`-sized batch
+                of `n x n` triangular matrices.
+            :attr:`upper` (bool):
+                If True, the tensor is upper-triangular, otherwise lower-triangular.
+        """
+        super().__init__(tensor)
+        if torch.is_tensor(tensor):
+            tensor = NonLazyTensor(tensor)
+        self._upper = upper
+        self._tensor = tensor
+
+    def __add__(self, other):
+        if isinstance(other, TriangularLazyTensor) and not self._upper ^ other.upper:
+            return self.__class__(self._tensor + other._tensor, upper=self._upper)
+        return self._tensor + other
+
+    @cached(name="cholesky")
+    def _cholesky(self):
+        return self.__class__(self._tensor._cholesky(), upper=self._upper)
+
+    def _mul_constant(self, constant):
+        return self.__class__(self._tensor * constant.unsqueeze(-1), upper=self._upper)
+
+    def _root_decomposition(self):
+        return self._tensor._root_decomposition()
+
+    def _root_inv_decomposition(self, initial_vectors=None):
+        self._tensor._root_inv_decomposition(initial_vectors=initial_vectors)
+
+    def _size(self):
+        return self._tensor.shape
+
+    def _sum_batch(self, dim):
+        return self.__class__(self._tensor._sum_batch(dim), upper=self._upper)
+
+    def abs(self):
+        return self.__class__(self._tensor.abs(), upper=self._upper)
+
+    def add_diag(self, added_diag):
+        shape = _mul_broadcast_shape(self._diag.shape, added_diag.shape)
+        return self.__class__(
+            AddedDiagLazyTensor(self._tensor.expand(shape), added_diag.expand(shape)), upper=self._upper
+        )
+
+    def diag(self):
+        return self._tensor.diag()
+
+    @cached
+    def evaluate(self):
+        return self._tensor.evaluate()
+
+    def exp(self):
+        return self.__class__(self._tensor.exp(), upper=self._upper)
+
+    def inverse(self):
+        eye = torch.eye(self._tensor.shape[-2:], device=self._tensor.device, dtype=self._tensor.dtype)
+        inv = self.inv_matmul(eye)
+        return self.__class__(inv, upper=self._upper)
+
+    def inv_matmul(self, right_tensor, left_tensor=None):
+        tsr = self.evaluate()
+        res = torch.triangular_solve(right_tensor, tsr, upper=self._upper).solution
+        if left_tensor is not None:
+            res = left_tensor @ res
+        return res
+
+    def inv_quad_logdet(self, inv_quad_rhs=None, logdet=False, reduce_inv_quad=True):
+        # triangular, inv_matmul is cheap
+        inv_quad_term = inv_quad_rhs.transpose(-1, -2) @ self.inv_matmul(inv_quad_rhs)
+        logdet_term = self.diag().prod(-1) if logdet else None
```

Comment:
> More serious addendum: if we want TLT to represent arbitrary triangular matrices and not just Cholesky roots, we'll also need to compute the sign of the determinant ahead of time, then do self.diag().abs().log().sum(-1) since we can't feed negative numbers in to the log.

So I guess the standing assumption throughout the lazies is psd. So if the det is negative we can return `NaN`, but that will likely mess with a lot of our assumptions. Maybe we should for now assume that things are psd and revisit the bigger picture of non-PSD lazies separately from this?